In [1]:
import numpy as np
import pandas as pd

In [2]:
train_all=pd.read_csv('train_all.csv')

In [3]:
train_lr=pd.read_csv('train_lr.csv')

In [4]:
train_dt=pd.read_csv('train_dt.csv')

In [5]:
train_all

inst_id  OC  sido_choongchung  sido_gangwon  sido_gyeongin  \
0          1   1               1.0           0.0            0.0   
1          3   1               0.0           0.0            0.0   
2          4   1               0.0           0.0            1.0   
3          7   1               0.0           0.0            1.0   
4          9   1               0.0           0.0            0.0   
..       ...  ..               ...           ...            ...   
296      422   1               0.0           0.0            0.0   
297      423   0               1.0           0.0            0.0   
298      426   0               0.0           0.0            0.0   
299      427   0               0.0           0.0            1.0   
300      428   1               0.0           0.0            0.0   

     sido_gyeongsang  sido_jeonra  sido_seoul  instkind_clinic  \
0                0.0          0.0         0.0              0.0   
1                1.0          0.0         0.0              0.0   
2                0.0          0.0         0.0              0.0   
3                0.0          0.0         0.0              0.0   
4                1.0          0.0         0.0              0.0   
..               ...          ...         ...              ...   
296              0.0          1.0         0.0              0.0   
297              0.0          0.0         0.0              0.0   
298              1.0          0.0         0.0              0.0   
299              0.0          0.0         0.0              0.0   
300              0.0          0.0         1.0              0.0   

     instkind_general_hospital  ...  liquidLiabilities2  shortLoan2  \
0                          0.0  ...           19.222130    0.000000   
1                          1.0  ...           18.841626    9.180719   
2                          0.0  ...            0.000000    0.000000   
3                          1.0  ...           23.557572   22.944579   
4                          1.0  ...           24.126880   23.590810   
..                         ...  ...                 ...         ...   
296                        0.0  ...           21.685753    0.000000   
297                        0.0  ...            0.000000    0.000000   
298                        0.0  ...           21.498441    0.000000   
299                        0.0  ...           22.579564    0.000000   
300                        0.0  ...           21.862083    0.000000   

     NCLiabilities2  longLoan2  netAsset2   surplus2  employee1  employee2  \
0         20.099863  19.781657  21.686169  20.963246   4.143135   4.174387   
1         18.812726  17.568559  18.838180  12.258481   6.687109   6.701960   
2          0.000000   0.000000   0.000000   0.000000   5.459586   0.693147   
3         23.755157  23.437961  23.284691  22.769775   6.498282   6.498282   
4         23.784786  23.370010  15.531458  22.923325   5.332719   5.288267   
..              ...        ...        ...        ...        ...        ...   
296       22.641007  22.503290  22.405862  20.514569   5.267858   4.955827   
297        0.000000   0.000000   0.000000   0.000000   4.382027   4.382027   
298       22.158142  21.639056  21.904736   0.000000   3.860891   3.905642   
299        0.000000   0.000000   0.000000   0.000000   4.615121   5.580909   
300       21.523440  21.507167  22.434570  22.558953   2.772589   2.772589   

     ownerChange  bedCount_class  
0            0.0             3.0  
1            0.0             4.0  
2            0.0             4.0  
3            0.0             4.0  
4            0.0             3.0  
..           ...             ...  
296          0.0             3.0  
297          0.0             3.2  
298          0.0             2.6  
299          0.0             3.0  
300          0.0             2.8  

[301 rows x 67 columns]

In [6]:
train_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 67 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   inst_id                        301 non-null    int64  
 1   OC                             301 non-null    int64  
 2   sido_choongchung               301 non-null    float64
 3   sido_gangwon                   301 non-null    float64
 4   sido_gyeongin                  301 non-null    float64
 5   sido_gyeongsang                301 non-null    float64
 6   sido_jeonra                    301 non-null    float64
 7   sido_seoul                     301 non-null    float64
 8   instkind_clinic                301 non-null    float64
 9   instkind_general_hospital      301 non-null    float64
 10  instkind_hospital              301 non-null    float64
 11  instkind_nursing_hospital      301 non-null    float64
 12  instkind_traditional_clinic    301 non-null    flo

In [7]:
train_all.set_index('inst_id',inplace=True)

In [8]:
train_lr.set_index('inst_id',inplace=True)
train_dt.set_index('inst_id',inplace=True)

In [9]:
train_all['ownerChange'].value_counts()

0.000000    257
1.000000     42
0.200000      1
0.145329      1
Name: ownerChange, dtype: int64

In [10]:
train_all['ownerChange']=round(train_all['ownerChange'])
train_all['ownerChange'].value_counts()

0.0    259
1.0     42
Name: ownerChange, dtype: int64

In [11]:
train_all['bedCount_class'].value_counts()

3.000000    163
2.000000     63
1.000000     26
4.000000     25
2.800000     10
3.200000      6
2.600000      4
3.400000      2
2.200000      1
2.671533      1
Name: bedCount_class, dtype: int64

In [12]:
train_all['bedCount_class']=round(train_all['bedCount_class'])
train_all['bedCount_class'].value_counts()

3.0    186
2.0     64
1.0     26
4.0     25
Name: bedCount_class, dtype: int64

In [13]:
X=train_all.drop('OC',axis=1)
Y=train_all['OC']

In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test=train_test_split(X,Y, random_state=23)

In [15]:
x_train, x_valid, y_train, y_valid=train_test_split(X_train,Y_train, random_state=4)

In [16]:
y_train.value_counts()

1    161
0      7
Name: OC, dtype: int64

In [17]:
y_valid.value_counts()

1    54
0     3
Name: OC, dtype: int64

In [18]:
Y_test.value_counts()

1    71
0     5
Name: OC, dtype: int64

In [19]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)
x_train_smote, y_train_smote = smote.fit_resample(x_train, y_train)

In [20]:
smote = SMOTE(random_state=42)
X_train_smote, Y_train_smote = smote.fit_resample(X_train, Y_train)

In [21]:
y_train_smote.value_counts()

1    161
0    161
Name: OC, dtype: int64

In [22]:
Y_train_smote.value_counts()

0    215
1    215
Name: OC, dtype: int64

In [23]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
train_features = scaler.fit_transform(x_train_smote)
test_features = scaler.transform(x_valid)

In [27]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
train_features = scaler.fit_transform(X_train_smote)
test_features = scaler.transform(X_test)

In [ ]:
from sklearn.neural_network import MLPClassifier

clf = MLPClassifier(solver = "lbfgs", alpha = 1e-5, hidden_layer_sizes = (5,2), random_state = 1)
clf.fit(X,y)

In [63]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import expon, reciprocal
from sklearn.model_selection import cross_val_score

param_distribs={
    'hidden_layer_sizes': [(10,),(20,),(30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,),
                           (10,10),(20,20),(30,30),(60,60),(10,30,10),(20,20,20),(20,40,60),(60,60,60)],
    'activation': ['identity', 'logistic', 'tanh', 'relu'],
    'solver': ['lbfgs', 'sgd', 'adam'],
#    'beta_1': reciprocal(0.1,1),
#    'beta_2': reciprocal(0.1,1),
#    'learning_rate_init': reciprocal(0.0001,0.1),
    'alpha': reciprocal(0.0001,0.1),
    'learning_rate': ['constant','invscaling','adaptive'],
}

mlp_gs = MLPClassifier(solver='lbfgs', early_stopping=True)

rnd_search=RandomizedSearchCV(mlp_gs, param_distributions=param_distribs, n_iter=50, cv=5, refit=True, scoring='accuracy', verbose=2)
rnd_search.fit(train_features, Y_train_smote)

Fitting 5 folds for each of 50 candidates, totalling 250 fits
[CV] END activation=relu, alpha=0.0002448536448270997, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.0002448536448270997, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.0002448536448270997, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.0002448536448270997, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.0002448536448270997, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=lbfgs; total time=   0.1s
[CV] END activation=relu, alpha=0.00017092578062381997, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=adam; total time=   0.0s
[CV] END activation=relu, alpha=0.00017092578062381997, hidden_layer_sizes=(70,), learning_rate=adaptive, solv

[CV] END activation=identity, alpha=0.02142425862824879, hidden_layer_sizes=(50,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.02142425862824879, hidden_layer_sizes=(50,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.02142425862824879, hidden_layer_sizes=(50,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.02142425862824879, hidden_layer_sizes=(50,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.02142425862824879, hidden_layer_sizes=(50,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0010968505033932747, hidden_layer_sizes=(20, 20), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0010968505033932747, hidden_layer_sizes=(20, 20), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activati

[CV] END activation=logistic, alpha=0.00020892422044722827, hidden_layer_sizes=(80,), learning_rate=constant, solver=lbfgs; total time=   0.0s
[CV] END activation=identity, alpha=0.007059541133462368, hidden_layer_sizes=(20,), learning_rate=constant, solver=lbfgs; total time=   0.2s
[CV] END activation=identity, alpha=0.007059541133462368, hidden_layer_sizes=(20,), learning_rate=constant, solver=lbfgs; total time=   0.0s
[CV] END activation=identity, alpha=0.007059541133462368, hidden_layer_sizes=(20,), learning_rate=constant, solver=lbfgs; total time=   0.1s
[CV] END activation=identity, alpha=0.007059541133462368, hidden_layer_sizes=(20,), learning_rate=constant, solver=lbfgs; total time=   0.0s
[CV] END activation=identity, alpha=0.007059541133462368, hidden_layer_sizes=(20,), learning_rate=constant, solver=lbfgs; total time=   0.1s
[CV] END activation=logistic, alpha=0.022633863524538958, hidden_layer_sizes=(60,), learning_rate=adaptive, solver=adam; total time=   0.0s
[CV] END act

[CV] END activation=identity, alpha=0.02866694420061788, hidden_layer_sizes=(70,), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.02866694420061788, hidden_layer_sizes=(70,), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.00038684860339102564, hidden_layer_sizes=(60,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.00038684860339102564, hidden_layer_sizes=(60,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.00038684860339102564, hidden_layer_sizes=(60,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.00038684860339102564, hidden_layer_sizes=(60,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END activation=identity, alpha=0.00038684860339102564, hidden_layer_sizes=(60,), learning_rate=constant, solver=adam; total time=   0.0s
[CV] END

C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=tanh, alpha=0.004556962520555645, hidden_layer_sizes=(20, 40, 60), learning_rate=constant, solver=lbfgs; total time=   1.3s
[CV] END activation=tanh, alpha=0.004556962520555645, hidden_layer_sizes=(20, 40, 60), learning_rate=constant, solver=lbfgs; total time=   0.0s
[CV] END activation=tanh, alpha=0.004556962520555645, hidden_layer_sizes=(20, 40, 60), learning_rate=constant, solver=lbfgs; total time=   0.4s
[CV] END activation=relu, alpha=0.00030329678526640506, hidden_layer_sizes=(80,), learning_rate=constant, solver=sgd; total time=   0.1s
[CV] END activation=relu, alpha=0.00030329678526640506, hidden_layer_sizes=(80,), learning_rate=constant, solver=sgd; total time=   0.1s
[CV] END activation=relu, alpha=0.00030329678526640506, hidden_layer_sizes=(80,), learning_rate=constant, solver=sgd; total time=   0.1s
[CV] END activation=relu, alpha=0.00030329678526640506, hidden_layer_sizes=(80,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activation=

C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=relu, alpha=0.04641913473286092, hidden_layer_sizes=(20,), learning_rate=adaptive, solver=lbfgs; total time=   0.3s
[CV] END activation=relu, alpha=0.04641913473286092, hidden_layer_sizes=(20,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.04641913473286092, hidden_layer_sizes=(20,), learning_rate=adaptive, solver=lbfgs; total time=   0.2s
[CV] END activation=logistic, alpha=0.00017912019554732565, hidden_layer_sizes=(20, 20), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.00017912019554732565, hidden_layer_sizes=(20, 20), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.00017912019554732565, hidden_layer_sizes=(20, 20), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.00017912019554732565, hidden_layer_sizes=(20, 20), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END

RandomizedSearchCV(cv=5,
                   estimator=MLPClassifier(early_stopping=True, solver='lbfgs'),
                   n_iter=50,
                   param_distributions={'activation': ['identity', 'logistic',
                                                       'tanh', 'relu'],
                                        'alpha': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001C2F7439FD0>,
                                        'hidden_layer_sizes': [(10,), (20,),
                                                               (30,), (40,),
                                                               (50,), (60,),
                                                               (70,), (80,),
                                                               (90,), (100,),
                                                               (10, 10),
                                                               (20, 20),
                                                               (30, 30),
                                                               (60, 60),
                                                               (10, 30, 10),
                                                               (20, 20, 20),
                                                               (20, 40, 60),
                                                               (60, 60, 60)],
                                        'learning_rate': ['constant',
                                                          'invscaling',
                                                          'adaptive'],
                                        'solver': ['lbfgs', 'sgd', 'adam']},
                   scoring='accuracy', verbose=2)

In [64]:
print(rnd_search.best_params_)
print(rnd_search.best_score_)

{'activation': 'relu', 'alpha': 0.0002448536448270997, 'hidden_layer_sizes': (60,), 'learning_rate': 'adaptive', 'solver': 'lbfgs'}
0.9720930232558139


In [65]:
pred=rnd_search.predict(train_features)
print(classificationSummary(Y_train_smote,pred))
rnd_search_pf = rnd_search.best_estimator_.predict(X_test)
print(classificationSummary(Y_test,rnd_search_pf))

Confusion Matrix (Accuracy 0.9977)

       Prediction
Actual   0   1
     0 215   0
     1   1 214
None
Confusion Matrix (Accuracy 0.9342)

       Prediction
Actual  0  1
     0  0  5
     1  0 71
None


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but MLPClassifier was fitted without feature names
  warnings.warn(


In [70]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import expon, reciprocal
from sklearn.model_selection import cross_val_score

param_distribs={
    'hidden_layer_sizes': [(10,),(20,),(30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,),
                           (10,10),(20,20),(30,30),(60,60),(10,20,10),(10,30,10),(20,20,20),(20,40,20),(60,60,60),(10,20,20,10),(20,40,40,20)],
    'activation': ['identity', 'logistic', 'tanh', 'relu'],
    'solver': ['lbfgs', 'sgd', 'adam'],
#    'beta_1': reciprocal(0.1,1),
#    'beta_2': reciprocal(0.1,1),
#    'learning_rate_init': reciprocal(0.0001,0.1),
    'alpha': reciprocal(0.0001,0.1),
    'learning_rate': ['constant','invscaling','adaptive'],
}

mlp_gs = MLPClassifier(solver='lbfgs', early_stopping=True)

rnd_search=RandomizedSearchCV(mlp_gs, param_distributions=param_distribs, n_iter=50, cv=5, refit=True, scoring='roc_auc', verbose=2)
rnd_search.fit(train_features, Y_train_smote)

Fitting 5 folds for each of 50 candidates, totalling 250 fits
[CV] END activation=relu, alpha=0.0014599400258276068, hidden_layer_sizes=(30,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=relu, alpha=0.0014599400258276068, hidden_layer_sizes=(30,), learning_rate=adaptive, solver=sgd; total time=   0.1s
[CV] END activation=relu, alpha=0.0014599400258276068, hidden_layer_sizes=(30,), learning_rate=adaptive, solver=sgd; total time=   0.1s
[CV] END activation=relu, alpha=0.0014599400258276068, hidden_layer_sizes=(30,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=relu, alpha=0.0014599400258276068, hidden_layer_sizes=(30,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=logistic, alpha=0.0018319147111729292, hidden_layer_sizes=(40,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=logistic, alpha=0.0018319147111729292, hidden_layer_sizes=(40,), learning_rate=adaptive, solver=sg

[CV] END activation=tanh, alpha=0.01671613683817949, hidden_layer_sizes=(40,), learning_rate=constant, solver=lbfgs; total time=   0.3s
[CV] END activation=identity, alpha=0.07158481115836791, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=identity, alpha=0.07158481115836791, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=identity, alpha=0.07158481115836791, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=sgd; total time=   0.2s
[CV] END activation=identity, alpha=0.07158481115836791, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=sgd; total time=   0.1s
[CV] END activation=identity, alpha=0.07158481115836791, hidden_layer_sizes=(70,), learning_rate=adaptive, solver=sgd; total time=   0.1s
[CV] END activation=identity, alpha=0.005008295498834376, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.3s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.005008295498834376, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.7s
[CV] END activation=identity, alpha=0.005008295498834376, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.4s
[CV] END activation=identity, alpha=0.005008295498834376, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.0s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.005008295498834376, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.7s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.0813485174780163, hidden_layer_sizes=(10, 10), learning_rate=adaptive, solver=lbfgs; total time=   0.2s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.0813485174780163, hidden_layer_sizes=(10, 10), learning_rate=adaptive, solver=lbfgs; total time=   0.2s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.0813485174780163, hidden_layer_sizes=(10, 10), learning_rate=adaptive, solver=lbfgs; total time=   0.2s
[CV] END activation=identity, alpha=0.0813485174780163, hidden_layer_sizes=(10, 10), learning_rate=adaptive, solver=lbfgs; total time=   0.0s


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.0813485174780163, hidden_layer_sizes=(10, 10), learning_rate=adaptive, solver=lbfgs; total time=   0.1s
[CV] END activation=logistic, alpha=0.0178326117518346, hidden_layer_sizes=(10, 30, 10), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0178326117518346, hidden_layer_sizes=(10, 30, 10), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0178326117518346, hidden_layer_sizes=(10, 30, 10), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0178326117518346, hidden_layer_sizes=(10, 30, 10), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=logistic, alpha=0.0178326117518346, hidden_layer_sizes=(10, 30, 10), learning_rate=invscaling, solver=adam; total time=   0.0s
[CV] END activation=tanh, alpha=0.06301273166465264, hidden_layer_sizes=(20, 20), learning_rate=invscaling, solver=sgd; tot

C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=identity, alpha=0.0031479840321133076, hidden_layer_sizes=(30, 30), learning_rate=adaptive, solver=lbfgs; total time=   0.3s
[CV] END activation=identity, alpha=0.014131993203101817, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=identity, alpha=0.014131993203101817, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=identity, alpha=0.014131993203101817, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=identity, alpha=0.014131993203101817, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=identity, alpha=0.014131993203101817, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.0936690485813454, hidden_layer_sizes=(90,), learning_rate=invscaling, solver=lbfgs; total time=   0.2s
[CV] END 

[CV] END activation=identity, alpha=0.00017120422111143702, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.0s
[CV] END activation=identity, alpha=0.00017120422111143702, hidden_layer_sizes=(60, 60), learning_rate=invscaling, solver=lbfgs; total time=   0.1s
[CV] END activation=logistic, alpha=0.01648966339090991, hidden_layer_sizes=(10,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.01648966339090991, hidden_layer_sizes=(10,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.01648966339090991, hidden_layer_sizes=(10,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.01648966339090991, hidden_layer_sizes=(10,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.01648966339090991, hidden_layer_sizes=(10,), learning_rate=constant, solver=sgd; total time=   0.0s
[CV] END activ

C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:559: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


[CV] END activation=relu, alpha=0.005357292498524926, hidden_layer_sizes=(10,), learning_rate=adaptive, solver=lbfgs; total time=   0.2s
[CV] END activation=relu, alpha=0.005357292498524926, hidden_layer_sizes=(10,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.005357292498524926, hidden_layer_sizes=(10,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=relu, alpha=0.005357292498524926, hidden_layer_sizes=(10,), learning_rate=adaptive, solver=lbfgs; total time=   0.0s
[CV] END activation=logistic, alpha=0.0005487252339661706, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.0005487252339661706, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logistic, alpha=0.0005487252339661706, hidden_layer_sizes=(50,), learning_rate=invscaling, solver=sgd; total time=   0.0s
[CV] END activation=logist

RandomizedSearchCV(cv=5,
                   estimator=MLPClassifier(early_stopping=True, solver='lbfgs'),
                   n_iter=50,
                   param_distributions={'activation': ['identity', 'logistic',
                                                       'tanh', 'relu'],
                                        'alpha': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001C2FA6556A0>,
                                        'hidden_layer_sizes': [(10,), (20,),
                                                               (30,), (40,),
                                                               (50,), (60,),
                                                               (70,), (80,),
                                                               (90,), (100,),
                                                               (10, 10),
                                                               (20, 20),
                                                               (30, 30),
                                                               (60, 60),
                                                               (10, 20, 10),
                                                               (10, 30, 10),
                                                               (20, 20, 20),
                                                               (20, 40, 20),
                                                               (60, 60, 60),
                                                               (10, 20, 20, 10),
                                                               (20, 40, 40,
                                                                20)],
                                        'learning_rate': ['constant',
                                                          'invscaling',
                                                          'adaptive'],
                                        'solver': ['lbfgs', 'sgd', 'adam']},
                   scoring='roc_auc', verbose=2)

In [71]:
print(rnd_search.best_params_)
print(rnd_search.best_score_)

{'activation': 'relu', 'alpha': 0.04817209417377487, 'hidden_layer_sizes': (10,), 'learning_rate': 'constant', 'solver': 'lbfgs'}
0.9908058409951325


In [72]:
pred=rnd_search.predict(train_features)
print(classificationSummary(Y_train_smote,pred))
rnd_search_pf = rnd_search.best_estimator_.predict(X_test)
print(classificationSummary(Y_test,rnd_search_pf))

Confusion Matrix (Accuracy 0.9977)

       Prediction
Actual   0   1
     0 215   0
     1   1 214
None
Confusion Matrix (Accuracy 0.9079)

       Prediction
Actual  0  1
     0  0  5
     1  2 69
None


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but MLPClassifier was fitted without feature names
  warnings.warn(


In [73]:
from sklearn.metrics import roc_auc_score
roc_auc_score(Y_test,rnd_search_pf)

0.4859154929577465

In [74]:
from sklearn.neural_network import MLPClassifier

mlp_gs = MLPClassifier(solver='lbfgs', early_stopping=True)
parameter_space = {
    'hidden_layer_sizes': [(10,),(20,),(30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,),
                           (10,10),(20,20),(30,30),(60,60),(10,20,10),(10,30,10),(20,20,20),(20,40,20),(60,60,60),(10,20,20,10),(20,40,40,20)],
#    'activation': ['identity', 'logistic', 'tanh', 'relu'],
#    'beta_1': [0.7,0.8,0.9],
#    'beta_2': [0.9,0.99,0.999],
#    'learning_rate_init': [0.0001,0.001,0.01],
    'alpha': [0.0001, 0.001, 0.01, 0.05],
    'learning_rate': ['constant','invscaling','adaptive']
}
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(mlp_gs, parameter_space, n_jobs=-1, cv=5, scoring='accuracy', verbose=2)
clf.fit(train_features, Y_train_smote) # X is train samples and y is the corresponding labels

Fitting 5 folds for each of 252 candidates, totalling 1260 fits


GridSearchCV(cv=5, estimator=MLPClassifier(early_stopping=True, solver='lbfgs'),
             n_jobs=-1,
             param_grid={'alpha': [0.0001, 0.001, 0.01, 0.05],
                         'hidden_layer_sizes': [(10,), (20,), (30,), (40,),
                                                (50,), (60,), (70,), (80,),
                                                (90,), (100,), (10, 10),
                                                (20, 20), (30, 30), (60, 60),
                                                (10, 20, 10), (10, 30, 10),
                                                (20, 20, 20), (20, 40, 20),
                                                (60, 60, 60), (10, 20, 20, 10),
                                                (20, 40, 40, 20)],
                         'learning_rate': ['constant', 'invscaling',
                                           'adaptive']},
             scoring='accuracy', verbose=2)

In [75]:
print('Best parameters found:\n', clf.best_params_)

Best parameters found:
 {'alpha': 0.001, 'hidden_layer_sizes': (30, 30), 'learning_rate': 'adaptive'}


In [77]:
pred=clf.predict(train_features)
print(classificationSummary(Y_train_smote,pred))
grid_search_pf = clf.best_estimator_.predict(X_test)
print(classificationSummary(Y_test,rnd_search_pf))

Confusion Matrix (Accuracy 0.9977)

       Prediction
Actual   0   1
     0 215   0
     1   1 214
None
Confusion Matrix (Accuracy 0.9079)

       Prediction
Actual  0  1
     0  0  5
     1  2 69
None


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but MLPClassifier was fitted without feature names
  warnings.warn(


In [78]:
from sklearn.metrics import roc_auc_score
roc_auc_score(Y_test,grid_search_pf)

0.5

In [79]:
from sklearn.neural_network import MLPClassifier

mlp_gs = MLPClassifier(solver='lbfgs', early_stopping=True)
parameter_space = {
    'hidden_layer_sizes': [(10,),(20,),(30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,),
                           (10,10),(20,20),(30,30),(60,60),(10,20,10),(10,30,10),(20,20,20),(20,40,20),(60,60,60),(10,20,20,10),(20,40,40,20)],
#    'activation': ['identity', 'logistic', 'tanh', 'relu'],
#    'beta_1': [0.7,0.8,0.9],
#    'beta_2': [0.9,0.99,0.999],
#    'learning_rate_init': [0.0001,0.001,0.01],
    'alpha': [0.0001, 0.001, 0.01, 0.05],
    'learning_rate': ['constant','invscaling','adaptive']
}
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(mlp_gs, parameter_space, n_jobs=-1, cv=5, scoring='roc_auc', verbose=1)
clf.fit(train_features, Y_train_smote) # X is train samples and y is the corresponding labels

Fitting 5 folds for each of 252 candidates, totalling 1260 fits


GridSearchCV(cv=5, estimator=MLPClassifier(early_stopping=True, solver='lbfgs'),
             n_jobs=-1,
             param_grid={'alpha': [0.0001, 0.001, 0.01, 0.05],
                         'hidden_layer_sizes': [(10,), (20,), (30,), (40,),
                                                (50,), (60,), (70,), (80,),
                                                (90,), (100,), (10, 10),
                                                (20, 20), (30, 30), (60, 60),
                                                (10, 20, 10), (10, 30, 10),
                                                (20, 20, 20), (20, 40, 20),
                                                (60, 60, 60), (10, 20, 20, 10),
                                                (20, 40, 40, 20)],
                         'learning_rate': ['constant', 'invscaling',
                                           'adaptive']},
             scoring='roc_auc', verbose=2)

In [80]:
print('Best parameters found:\n', clf.best_params_)

Best parameters found:
 {'alpha': 0.0001, 'hidden_layer_sizes': (10, 30, 10), 'learning_rate': 'adaptive'}


In [81]:
pred=clf.predict(train_features)
print(classificationSummary(Y_train_smote,pred))
grid_search_pf = clf.best_estimator_.predict(X_test)
print(classificationSummary(Y_test,rnd_search_pf))

Confusion Matrix (Accuracy 0.9977)

       Prediction
Actual   0   1
     0 215   0
     1   1 214
None
Confusion Matrix (Accuracy 0.9079)

       Prediction
Actual  0  1
     0  0  5
     1  2 69
None


C:\Users\ynjn\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but MLPClassifier was fitted without feature names
  warnings.warn(


In [83]:
roc_auc_score(Y_train_smote,pred)

0.9976744186046511

In [82]:
from sklearn.metrics import roc_auc_score
roc_auc_score(Y_test,grid_search_pf)

0.5

# 모델 성능 정리
  
- SVM  
train 정확도 : 0.9977, AUC: 0.9977  
test 정확도 : 0.9342, AUC : 0.593  
- 판별분석  
    - LDA  
train 정확도 : 0.9628, AUC: 0.9628  
test 정확도 : 0.8816, AUC : 0.4728  
    -QDA  
train 정확도 : 0.9977, AUC: 0.9977  
test 정확도 : 0.9342, AUC : 0.593  
-KNeighborsClassifier(컷오프 0.4)  
train 정확도 : 0.9977, AUC: 0.9977  
test 정확도 : 0.8684, AUC : 0.6507
-SVM&KNeighborsClassifier Ensemble  
train 정확도 : 0.9977, AUC: 0.9977  
test 정확도 : 0.9211, AUC : 0.586  
-neural network  
train 정확도 : 0.9977, AUC: 0.9977  
test 정확도 : 0.9079, AUC : 0.5  